src/pmdas/uwsgi/pmdauwsgi.python

```diff
@@ -0,0 +1,111 @@
+# option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+# for more details.
+#
+
+# pylint: disable=line-too-long, too-many-return-statements
+# pylint: disable=too-many-statements, too-many-branches, too-many-lines
+# pylint: disable=broad-except, too-many-instance-attributes
+
+""" Performance Metrics Domain Agent exporting uwsgi metrics. """
+
+from cpmapi import (
+    PM_TYPE_STRING,
+    PM_INDOM_NULL,
+    PM_SEM_INSTANT,
+    PM_TYPE_DOUBLE,
+    PM_TYPE_U64,
+    PM_SEM_COUNTER,
+    PM_ERR_PMID,
+    PM_ERR_AGAIN
+)
+
+from pcp.pmapi import pmUnits, pmContext
+from pcp.pmda import PMDA, pmdaMetric
+
+import os
+import requests
+
+
+def _is_pmda_setup():
+    """checks if PMDA is in setup state"""
+    return os.environ.get('PCP_PYTHON_DOMAIN') or os.environ.get('PCP_PYTHON_PMNS')
+
+
+class UwsgiPMDA(PMDA):
+
+    def __init__(self, name, domain):
+        ''' Initialisation - register metrics, callbacks, drop privileges '''
+
+        PMDA.__init__(self, name, domain)
+        self.connect_pmcd()
+
+        self.add_metric(name + '.workers', pmdaMetric(PMDA.pmid(0, 0), PM_TYPE_U64, PM_INDOM_NULL, PM_SEM_COUNTER, pmUnits()))
+        self.add_metric(name + '.avg_response_time_msec', pmdaMetric(PMDA.pmid(0, 1), PM_TYPE_DOUBLE, PM_INDOM_NULL, PM_SEM_COUNTER, pmUnits()))
+        self.add_metric(name + '.total_requests_served', pmdaMetric(PMDA.pmid(0, 2), PM_TYPE_U64, PM_INDOM_NULL, PM_SEM_COUNTER, pmUnits()))
+        self.add_metric(name + '.workers_accepting_requests', pmdaMetric(PMDA.pmid(0, 3), PM_TYPE_U64, PM_INDOM_NULL, PM_SEM_COUNTER, pmUnits()))
+        self.add_metric(name + '.total_exceptions', pmdaMetric(PMDA.pmid(0, 4), PM_TYPE_U64, PM_INDOM_NULL, PM_SEM_COUNTER, pmUnits()))
+        self.add_metric(name + '.total_harakiri_count', pmdaMetric(PMDA.pmid(0, 5), PM_TYPE_U64, PM_INDOM_NULL, PM_SEM_COUNTER, pmUnits()))
+        self.add_metric(name + '.busy_worker_count', pmdaMetric(PMDA.pmid(0, 6), PM_TYPE_U64, PM_INDOM_NULL, PM_SEM_COUNTER, pmUnits()))
+        self.add_metric(name + '.idle_worker_count', pmdaMetric(PMDA.pmid(0, 7), PM_TYPE_U64, PM_INDOM_NULL, PM_SEM_COUNTER, pmUnits()))
+        self.add_metric(name + '.pause_worker_count', pmdaMetric(PMDA.pmid(0, 8), PM_TYPE_U64, PM_INDOM_NULL, PM_SEM_COUNTER, pmUnits()))
+
+        self.set_fetch_callback(self.uwsgi_fetch_callback)
+        self.set_user(pmContext.pmGetConfig('PCP_USER'))
+
+    def uwsgi_fetch_callback(self, cluster, item, inst):
+        ''' Returns a list of value,status (single pair) for one metric '''
+
+        if not cluster == 0:
+            return [PM_ERR_PMID, 0]
+        elif not _is_pmda_setup():
+            try:
+                stats = requests.get('http://127.0.0.1:9051').json()
```

Comment:
It might be best if this URL (especially the port) came from a configuration file?  See pmdalibvirt.python as one common way this is done elsewhere (configparser module).